src/main/java/ch/njol/skript/effects/EffTeleport.java

```diff
@@ -127,34 +137,43 @@ protected TriggerItem walk(Event e) {
 
 		Delay.addDelayedEvent(e);
 		Object localVars = Variables.removeLocals(e);
-		
+
 		// This will either fetch the chunk instantly if on Spigot or already loaded or fetch it async if on Paper.
-		PaperLib.getChunkAtAsync(loc).thenAccept(chunk -> {
-			// The following is now on the main thread
-			for (Entity entity : entityArray) {
-				EntityUtils.teleport(entity, loc);
-			}
+		CompletableFuture<Chunk> chunkFuture = PaperLib.getChunkAtAsync(loc).thenApply(chunk -> {
+			Bukkit.getScheduler().runTask(Skript.getInstance(), () -> {
```

Comment:
might be worth using the Task util class, e.g. https://github.com/SkriptLang/Skript/blob/0bfec0b709cf74bace063d315d0eb501058cac14/src/main/java/ch/njol/skript/util/Task.java#L147